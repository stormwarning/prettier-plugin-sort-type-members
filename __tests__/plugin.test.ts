import { readdir, readFile } from 'node:fs/promises'
import { join } from 'node:path'

import { ESLint } from 'eslint'
import { format } from 'prettier'
import { describe, expect, test } from 'vitest'

const plugins = ['./dist/index.js']

const fixtures = join(import.meta.dirname, 'fixtures')
const filenames = await readdir(fixtures)
const options = [{}, { sortMembersAlphabetically: true }, { sortMembersAlphabetically: false }]

describe.each(options)('format, %j', (_options) => {
	test.each(filenames)('%s', async (name) => {
		let path = join(fixtures, name)
		let code = await readFile(path, 'utf8')
		let result = await format(code, {
			..._options,
			filepath: path,
			plugins,
		})

		expect(result).toMatchSnapshot()
	})

	describe('idempotency', () => {
		test.each(filenames)('%s', async (name) => {
			let path = join(fixtures, name)
			let code = await readFile(path, 'utf8')
			let result1 = await format(code, {
				..._options,
				filepath: path,
				plugins,
			})
			let result2 = await format(result1, {
				..._options,
				filepath: path,
				plugins,
			})

			expect(result2).toEqual(result1)
		})
	})

	//
	// describe('parser agnostic', () => {
	// 	describe('TypeScript', () => {
	// 		let skipFile = new Set(['issue-34-literal-keys.js'])
	// 		let parsers = ['typescript', 'babel-ts']
	// 		describe.each(parsers)('%s', async (parser) => {
	// 			test.each(filenames.filter((n) => !skipFile.has(n)))('%s', async (name) => {
	// 				let path = join(dir, name)
	// 				let code = await readFile(path, 'utf-8')
	// 				let expected = await format(code, {
	// 					..._options,
	// 					filepath: path,
	// 					plugins,
	// 				})
	// 				let actual = await format(code, {
	// 					..._options,
	// 					filepath: path,
	// 					plugins,
	// 					parser,
	// 				})
	// 				expect(actual).toBe(expected)
	// 			})
	// 		})
	// 	})
	// })

	// describe('compatible with eslint-typescript', () => {
	// 	let eslint = new ESLint({
	// 		overrideConfig: {
	// 			parser: '@typescript-eslint/parser',
	// 			plugins: ['@typescript-eslint'],
	// 			extends: [],
	// 			rules: {
	// 				'@typescript-eslint/member-ordering': 'error',
	// 			},
	// 		},
	// 		useEslintrc: false,
	// 	})

	// 	test.each(filenames)('%s', async (name) => {
	// 		let path = join(dir, name)
	// 		let code = await readFile(path, 'utf-8')
	// 		let result = await format(code, {
	// 			..._options,
	// 			filepath: path,
	// 			plugins,
	// 		})

	// 		let lintResults = await eslint.lintText(result)

	// 		expect(lintResults).toHaveLength(1)
	// 		expect(lintResults[0].messages).toBeEmpty()
	// 	})
	// })
})
