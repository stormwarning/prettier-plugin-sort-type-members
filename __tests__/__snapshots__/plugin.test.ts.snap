// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`format, {"sortMembersAlphabetically":false} > function-member.ts 1`] = `
"type Link = {
  foo: () => void;
  text: string;
  onClick: () => void;
};

type Literal = {
  method(): void;
  func: () => void;
  value: number;
};

interface Interface {
  method(): void;
  func: () => void;
  value: number;
}
"
`;

exports[`format, {"sortMembersAlphabetically":false} > interface.ts 1`] = `
"interface A {
  b: string;
  c: unknown;
  a: string;
}

interface B {
  x: string;
  a(): string;
  z(): string;
  b(): never;
  c: number;
}

interface C {
  z: 0;
  a: string;

  b: string;
  aa: 0;
  m(): unknown;

  c: string;
}

interface D {
  a: string;
  b: string;

  c: 0;

  am(): 0;
  bm(): 0;
}

interface E {
  // comment-on-a
  a: 0;

  // comment-on-c
  c: 0; // tail-comment-c

  // comment-on-d
  d(): 0; // tail-comment-d
  b: 0; // tail-comment-b

  e(): 0;
  // trailing-comment
}

interface F {
  a(): 0;
  b: 0;
  [c: number]: 0;
  new (): F;
  new (a: 0, b: 0): F;
  [d: symbol]: 0;
  [\`e\`]: 0;
  [\`f\`]: 0;
  new (a: 0): F;
  [\`g\`](): 0;
  set h(_: 0);
  get i(): 0;
  j: () => 0;
}

interface G {
  a: 0;
  [c: number]: () => 0;
  [d: symbol]: 0;
}
"
`;

exports[`format, {"sortMembersAlphabetically":false} > literal-keys.ts 1`] = `
"type Endpoint<T> = T | unknown;
type UserResponse = unknown;
type RepairOrdersResponse = unknown;

export type DataEndpoints = {
  user: Endpoint<UserResponse>;
  "repair-orders": Endpoint<RepairOrdersResponse>;
};

type X = {
  data: unknown;
  "background-color": unknown;
  ["color"]: unknown;
  alpha: unknown;
};

const alpha = Symbol();
type Y = {
  date: Date;
  best: unknown;
  ["circle"]: unknown;
  [alpha]: unknown;
  alpha: unknown;
};

type Z = {
  edge: () => void;
  dog: unknown;
  "change-value": () => void;
  "big-value": number;
  "another-one": 0;
};

interface A {
  [alpha]: unknown;
  one: unknown;
  ["two"]: unknown;
  ["three"](): void;
  four: unknown;
  five(): void;
}
"
`;

exports[`format, {"sortMembersAlphabetically":false} > type-alias.ts 1`] = `
"type AType = {
  z(): number;
  y: number;
  x: () => () => number;
  new (a: 0, b: 1): A;
  get w(): number;
  v?(): number;
  [u: number]: () => number;
  [t: string]: unknown;
  [\`s\`]: number;
  r?: number;
};
"
`;

exports[`format, {"sortMembersAlphabetically":true} > function-member.ts 1`] = `
"type Link = {
  foo: () => void;
  text: string;
  onClick: () => void;
};

type Literal = {
  method(): void;
  func: () => void;
  value: number;
};

interface Interface {
  method(): void;
  func: () => void;
  value: number;
}
"
`;

exports[`format, {"sortMembersAlphabetically":true} > interface.ts 1`] = `
"interface A {
  b: string;
  c: unknown;
  a: string;
}

interface B {
  x: string;
  a(): string;
  z(): string;
  b(): never;
  c: number;
}

interface C {
  z: 0;
  a: string;

  b: string;
  aa: 0;
  m(): unknown;

  c: string;
}

interface D {
  a: string;
  b: string;

  c: 0;

  am(): 0;
  bm(): 0;
}

interface E {
  // comment-on-a
  a: 0;

  // comment-on-c
  c: 0; // tail-comment-c

  // comment-on-d
  d(): 0; // tail-comment-d
  b: 0; // tail-comment-b

  e(): 0;
  // trailing-comment
}

interface F {
  a(): 0;
  b: 0;
  [c: number]: 0;
  new (): F;
  new (a: 0, b: 0): F;
  [d: symbol]: 0;
  [\`e\`]: 0;
  [\`f\`]: 0;
  new (a: 0): F;
  [\`g\`](): 0;
  set h(_: 0);
  get i(): 0;
  j: () => 0;
}

interface G {
  a: 0;
  [c: number]: () => 0;
  [d: symbol]: 0;
}
"
`;

exports[`format, {"sortMembersAlphabetically":true} > literal-keys.ts 1`] = `
"type Endpoint<T> = T | unknown;
type UserResponse = unknown;
type RepairOrdersResponse = unknown;

export type DataEndpoints = {
  user: Endpoint<UserResponse>;
  "repair-orders": Endpoint<RepairOrdersResponse>;
};

type X = {
  data: unknown;
  "background-color": unknown;
  ["color"]: unknown;
  alpha: unknown;
};

const alpha = Symbol();
type Y = {
  date: Date;
  best: unknown;
  ["circle"]: unknown;
  [alpha]: unknown;
  alpha: unknown;
};

type Z = {
  edge: () => void;
  dog: unknown;
  "change-value": () => void;
  "big-value": number;
  "another-one": 0;
};

interface A {
  [alpha]: unknown;
  one: unknown;
  ["two"]: unknown;
  ["three"](): void;
  four: unknown;
  five(): void;
}
"
`;

exports[`format, {"sortMembersAlphabetically":true} > type-alias.ts 1`] = `
"type AType = {
  z(): number;
  y: number;
  x: () => () => number;
  new (a: 0, b: 1): A;
  get w(): number;
  v?(): number;
  [u: number]: () => number;
  [t: string]: unknown;
  [\`s\`]: number;
  r?: number;
};
"
`;

exports[`format, {} > function-member.ts 1`] = `
"type Link = {
  foo: () => void;
  text: string;
  onClick: () => void;
};

type Literal = {
  method(): void;
  func: () => void;
  value: number;
};

interface Interface {
  method(): void;
  func: () => void;
  value: number;
}
"
`;

exports[`format, {} > interface.ts 1`] = `
"interface A {
  b: string;
  c: unknown;
  a: string;
}

interface B {
  x: string;
  a(): string;
  z(): string;
  b(): never;
  c: number;
}

interface C {
  z: 0;
  a: string;

  b: string;
  aa: 0;
  m(): unknown;

  c: string;
}

interface D {
  a: string;
  b: string;

  c: 0;

  am(): 0;
  bm(): 0;
}

interface E {
  // comment-on-a
  a: 0;

  // comment-on-c
  c: 0; // tail-comment-c

  // comment-on-d
  d(): 0; // tail-comment-d
  b: 0; // tail-comment-b

  e(): 0;
  // trailing-comment
}

interface F {
  a(): 0;
  b: 0;
  [c: number]: 0;
  new (): F;
  new (a: 0, b: 0): F;
  [d: symbol]: 0;
  [\`e\`]: 0;
  [\`f\`]: 0;
  new (a: 0): F;
  [\`g\`](): 0;
  set h(_: 0);
  get i(): 0;
  j: () => 0;
}

interface G {
  a: 0;
  [c: number]: () => 0;
  [d: symbol]: 0;
}
"
`;

exports[`format, {} > literal-keys.ts 1`] = `
"type Endpoint<T> = T | unknown;
type UserResponse = unknown;
type RepairOrdersResponse = unknown;

export type DataEndpoints = {
  user: Endpoint<UserResponse>;
  "repair-orders": Endpoint<RepairOrdersResponse>;
};

type X = {
  data: unknown;
  "background-color": unknown;
  ["color"]: unknown;
  alpha: unknown;
};

const alpha = Symbol();
type Y = {
  date: Date;
  best: unknown;
  ["circle"]: unknown;
  [alpha]: unknown;
  alpha: unknown;
};

type Z = {
  edge: () => void;
  dog: unknown;
  "change-value": () => void;
  "big-value": number;
  "another-one": 0;
};

interface A {
  [alpha]: unknown;
  one: unknown;
  ["two"]: unknown;
  ["three"](): void;
  four: unknown;
  five(): void;
}
"
`;

exports[`format, {} > type-alias.ts 1`] = `
"type AType = {
  z(): number;
  y: number;
  x: () => () => number;
  new (a: 0, b: 1): A;
  get w(): number;
  v?(): number;
  [u: number]: () => number;
  [t: string]: unknown;
  [\`s\`]: number;
  r?: number;
};
"
`;
